sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -227,35 +230,37 @@ static void CleanupThread()
       // of this, the oldest connection in the pool can be found at the end of the list. Looping the
       // connection pool backwards until a connection that is not expired is found or until all
       // connections are removed.
-      for (auto connection = --(index->second.end());
-           index->second.size() > 0 && connection->get()->IsExpired();
-           connection = index->second.size() > 0 ? --connection : connection)
+      auto& connectionList = index->second;
+      auto connectionIter = connectionList.end();
+      while (connectionIter != connectionList.begin())
       {
-        // remove connection from the pool and update the connection to the next one
-        // which is going to be list.end()
-        Log::Write(Logger::Level::Verbose, "Clean pool - remove connection");
-        connection = index->second.erase(connection);
+        --connectionIter;
+        if ((*connectionIter)->IsExpired())
+        {
+          // remove connection from the pool and update the connection to the next one
+          // which is going to be list.end()
+          connectionsToBeCleaned.emplace_back(std::move(*connectionIter));
+          connectionIter = connectionList.erase(connectionIter);
+        }
+        else
+        {
+          break;
+        }
       }
 
-      if (index->second.size() == 0)
+      if (connectionList.empty())
       {
         Log::Write(Logger::Level::Verbose, "Clean pool - remove index " + index->first);
         index = CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.erase(index);
       }
       else
       {
-        index = ++index;
+        ++index;
       }
     }
 
-    if (CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.size() == 0)
-    {
-      Log::Write(
-          Logger::Level::Verbose,
-          "Clean pool - all connections removed. Return**********************");
-      CurlConnectionPool::g_curlConnectionPool.IsCleanThreadRunning = false;
-      return;
-    }
+    lockForPoolCleaning.unlock();
+    // Do actual connections release work here, without holding the mutex.
```

Comment:
Right, so, we were calling curl_easy_cleanup for each expired connection while holding the lock?  And eventually if the clean routine takes too long, new connections can't be created.  Great finding @Jinming-Hu ! thank you!